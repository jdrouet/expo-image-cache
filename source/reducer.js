import { createAction, handleActions } from 'redux-actions';

export const constants = {
  init: 'expo-image-cache.init',
  initSuccess: 'expo-image-cache.init-success',
  initError: 'expo-image-cache.init-error',
  resolve: 'expo-image-cache.resolve',
  download: 'expo-image-cache.download',
  downloadSuccess: 'expo-image-cache.download-success',
  downloadError: 'expo-image-cache.download-error',
};

export const actions = Object
  .keys(constants)
  .reduce((res, key) => Object.assign(res, {
    [key]: createAction(constants[key]),
  }), {});

export const initState = {
  initializing: false,
  initialized: false,
  downloading: [],
  downloaded: {},
};

export default handleActions({
  [constants.init]: (state) => ({
    ...state,
    initializing: true,
  }),
  [constants.initSuccess]: (state) => ({
    ...state,
    initializing: false,
    initialized: true,
  }),
  [constants.initError]: (state) => ({
    ...state,
    initializing: false,
    initialized: false,
  }),
  [constants.download]: (state, { payload }) => ({
    ...state,
    downloading: [
      ...state.downloading,
      payload.uri,
    ],
  }),
  [constants.downloadSuccess]: (state, { payload }) => {
    const downloading = state.downloading.filter((uri) => uri !== payload.uri);
    const downloaded = {
      ...state.downloaded,
      [payload.uri]: payload.local,
    };
    return {
      ...state,
      downloading,
      downloaded,
    };
  },
  [constants.downloadError]: (state, { payload }) => {
    const downloading = state.downloading.filter((uri) => uri !== payload.uri);
    return {
      ...state,
      downloading,
    };
  },
}, initState);
