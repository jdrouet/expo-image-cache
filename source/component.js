import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Image } from 'react-native';
import { actions } from './reducer';

class SmartImage extends React.Component {
  static propTypes = {
    component: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.func,
    ]),
    source: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.shape({
        uri: PropTypes.string.isRequired,
      }),
    ]).isRequired,
    imageCacheAction: PropTypes.shape({
      resolve: PropTypes.func.isRequired,
    }).isRequired,
    file: PropTypes.string,
    loading: PropTypes.bool.isRequired,
    remote: PropTypes.bool.isRequired,
    renderLoading: PropTypes.func,
  };

  static defaultProps = {
    component: Image,
    file: null,
    renderLoading: null,
  };

  componentWillMount() {
    if (this.props.remote) {
      this.props.imageCacheAction.resolve(this.props.source);
    }
  }

  shouldComponentUpdate(nextProps) {
    return (
      this.props.remote
      && this.props.file !== nextProps.file
    ) || (
      this.props.source !== nextProps.source
    );
  }

  getSource() {
    if (!this.props.remote || this.props.loading) return this.props.source;
    return { uri: this.props.file };
  }

  render() {
    if (this.props.loading && this.props.renderLoading) {
      return this.props.renderLoading(this.props);
    }
    const Component = this.props.component;
    const source = this.getSource();
    return (
      <Component
        {...this.props}
        source={source}
      />
    );
  }
}

export const Component = SmartImage;

export const isRemote = (props) =>
  !!props.source && !!props.source.uri;

// TODO find a way to modify the reducer location
export const isLoading = (state, params) => {
  if (!isRemote(params)) return false;
  const list = state.imageCache.downloading;
  return list.includes(params.source.uri);
};

export const getFile = (state, params) => {
  if (!isRemote(params) || isLoading(state, params)) return null;
  return state.imageCache.downloaded[params.source.uri];
};

export const paramsToProps = (state, params) => ({
  remote: isRemote(params),
  loading: isLoading(state, params),
  file: getFile(state, params),
});

export const bindActionsToProps = (dispatch) => ({
  imageCacheAction: bindActionCreators(actions, dispatch),
});

export default connect(
  paramsToProps,
  bindActionsToProps,
)(SmartImage);
