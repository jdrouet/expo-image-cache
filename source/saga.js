import { FileSystem } from 'expo';
import sha256 from 'crypto-js/sha256';
import { put, select, takeEvery, takeLatest } from 'redux-saga/effects';
import { actions, constants } from './reducer';

export const resolveFilename = (uri) => sha256(uri).toString();

export const defaultConfig = {
  folderName: 'expo-image-cache',
  reducerName: 'imageCache',
  resolveFilename,
};

export const getLoading = (config) => (state) =>
  state[config.reducerName].loading;

export default (givenConfig = {}) => {
  const config = Object.assign({}, defaultConfig, givenConfig);
  const folder = FileSystem.cacheDirectory + config.folderName;

  const getLoadingInstance = getLoading(config);

  const resolve = function* ({ payload }) {
    const loading = yield select(getLoadingInstance);
    if (loading.includes(payload.uri)) return;
    const filename = config.resolveFilename(payload);
    const file = `${folder}/${filename}`;
    const localFile = yield FileSystem.getInfoAsync(file);
    if (localFile.exists) {
      yield put(actions.downloadSuccess({
        uri: payload.uri,
        local: localFile.uri,
      }));
      return;
    }
    yield put(actions.download(payload.uri));
    try {
      const downloaded = yield FileSystem.downloadAsync(payload.uri, file);
      yield put(actions.downloadSuccess({
        uri: payload.uri,
        local: downloaded.uri,
      }));
    } catch (error) {
      yield put(actions.downloadError(error));
    }
  };

  const resolveSaga = function* () {
    yield takeEvery(constants.resolve, resolve);
  };

  const init = function* () {
    const folderInfo = yield FileSystem.getInfoAsync(folder);
    if (folderInfo.exists) return;
    try {
      yield FileSystem.makeDirectoryAsync(folder);
      yield put(actions.initSuccess());
    } catch (error) {
      yield put(actions.initError(error));
    }
  };

  const initSaga = function* () {
    yield takeLatest(constants.init, init);
  };

  return {
    config,
    folder,
    getLoading: getLoadingInstance,
    init,
    initSaga,
    resolve,
    resolveSaga,
  };
};
