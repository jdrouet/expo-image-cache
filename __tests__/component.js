import 'react-native';
import React from 'react';
import { Component, getFile, isLoading, isRemote } from '../source/component';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

import { actions } from '../source/reducer';

const uri = 'http://foo.bar/image.png';

describe('component', () => {
  describe('getFile', () => {
    it('should detect local files', () => {
      expect(isRemote({source: {uri}})).toEqual(true);
    })
  });

  describe('isLoading', () => {
    it('should return true', () => {
      expect(isLoading({
        imageCache: {
          downloading: [uri],
        },
      }, {
        source: {uri},
      })).toEqual(true);
    })

    it('should return false', () => {
      expect(isLoading({
        imageCache: {
          downloading: [],
        },
      }, {
        source: {uri},
      })).toEqual(false);
    })
  });

  describe('getFile', () => {
    it('should detect local files', () => {
      expect(getFile({
        imageCache: {
          downloading: [],
          downloaded: {
            [uri]: 'here',
          },
        },
      }, {
        source: {uri},
      })).toEqual('here');
    })
  });

  describe('render', () => {
    it('should render the remote file', () => {
      const tree = renderer.create(
        <Component
          imageCacheAction={actions}
          loading={false}
          remote={false}
          source={4}
        />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render the loading function', () => {
      const renderLoading = () => null;
      const source = {uri};
      const tree = renderer.create(
        <Component
          imageCacheAction={actions}
          loading={true}
          remote={true}
          source={source}
          renderLoading={renderLoading}
        />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render the image even if loading', () => {
      const source = {uri};
      const tree = renderer.create(
        <Component
          imageCacheAction={actions}
          loading={true}
          remote={true}
          source={source}
        />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render the local file', () => {
      const source = {uri};
      const file = 'file://local/image.png';
      const tree = renderer.create(
        <Component
          imageCacheAction={actions}
          loading={false}
          remote={true}
          source={source}
          file={file}
        />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
