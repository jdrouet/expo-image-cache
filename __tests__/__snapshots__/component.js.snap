// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component render should render the image even if loading 1`] = `
<Image
  component={[Function]}
  file={null}
  imageCacheAction={
    Object {
      "download": [Function],
      "downloadError": [Function],
      "downloadSuccess": [Function],
      "init": [Function],
      "initError": [Function],
      "initSuccess": [Function],
      "resolve": [Function],
    }
  }
  loading={true}
  remote={true}
  renderLoading={null}
  source={
    Object {
      "uri": "http://foo.bar/image.png",
    }
  }
/>
`;

exports[`component render should render the loading function 1`] = `null`;

exports[`component render should render the local file 1`] = `
<Image
  component={[Function]}
  file="file://local/image.png"
  imageCacheAction={
    Object {
      "download": [Function],
      "downloadError": [Function],
      "downloadSuccess": [Function],
      "init": [Function],
      "initError": [Function],
      "initSuccess": [Function],
      "resolve": [Function],
    }
  }
  loading={false}
  remote={true}
  renderLoading={null}
  source={
    Object {
      "uri": "file://local/image.png",
    }
  }
/>
`;

exports[`component render should render the remote file 1`] = `
<Image
  component={[Function]}
  file={null}
  imageCacheAction={
    Object {
      "download": [Function],
      "downloadError": [Function],
      "downloadSuccess": [Function],
      "init": [Function],
      "initError": [Function],
      "initSuccess": [Function],
      "resolve": [Function],
    }
  }
  loading={false}
  remote={false}
  renderLoading={null}
  source={4}
/>
`;
